crates/api_server/src/server.rs

```diff
@@ -127,3 +146,33 @@ impl NodeServer {
         (self.run_fn)()
     }
 }
+
+#[derive(Clone)]
+pub struct TelemetryReporter<S> {
+    service: S,
+    telemetry: &'static Telemetry,
+}
+
+impl<'a, S> RpcServiceT<'a> for TelemetryReporter<S>
+where
+    S: RpcServiceT<'a> + Send + Sync + Clone + 'static,
+{
+    type Future = BoxFuture<'a, MethodResponse>;
+
+    fn call(&self, req: Request<'a>) -> Self::Future {
+        let service = self.service.clone();
+        let telemetry = self.telemetry; // Capture telemetry reference
+
+        async move {
+            let method = req.method_name();
+            // Report only anvil and config API usage
+            if method.starts_with("anvil_") || method.starts_with("config_") {
+                let mut event_props = HashMap::new();
+                event_props.insert("method".to_string(), method.into());
+                let _ = telemetry.track_event("rpc_call", event_props).await;
```

Comment:
BTW with proposed `TelemetryProps` you can improve the ergonomics even more if you will also add this:

```rust
impl<V: ToString> From<(&'static str, V)> for TelemetryProps {
  fn from(kv: ('static str, V)) -> Self {
    let mut self_ = Self::default();
    self.insert(kv.0, kv.1);
    self
  }
}

impl From<TelemetryProps> for HashMap<String, String> {
  fn from(props: TelemetryProps) -> Self {
    props.into_inner()
  }
}
```

...and if `track_event` will accept `T: Into<HashMap<String, String>>` (or `TelemetryProps` will be a part of the interface there), then this snipped can be simplified to:

```rust
telemetry.track_event("rpc_call", TelemetryProps::from(("method", method)).await.ok();
```

Reply:
Love it, what do you think about using `serde_json::Value` though? This way we can do something like:

`TelemetryProps::from(json!({ "method": method, "some_other_prop": some_other_value }))`

Reply:
Instead added method to finalize the builder, this way any props can be initialized in a single line.